File: force-field/ruby/Vector.rb
class Vector
  attr_accessor :x
  attr_accessor :y
  
  def initialize(x, y)
    @x = x
    @y = y
  end
  
  def subtract(v)
    Vector.new @x - v.x, @y - v.y
  end
  
  def add(v)
    Vector.new @x + v.x, @y + v.y
  end
  
  def scale(s)
    Vector.new x * s, y * s
  end
  
  def addInPlace(v)
    @x += v.x
    @y += v.y
  end
  
  def dot(v)
    @x * v.x + @y * v.y
  end
  
  def normalize()
    magnitude = Math.sqrt self.dot self
    if magnitude == 0
      return Vector.new 0, 0
    end
    self.scale 1.0 / magnitude
  end
  
  def to_s
    "(#{@x}, #{@y})"
  end
end

File: force-field/python/vector.py
import math

class Vector:
  def __init__(self, x, y):
    self.x = x
    self.y = y
  
  def subtract(self, v):
    return Vector(self.x - v.x, self.y - v.y);
  
  def add(self, v):
    return Vector(self.x + v.x, self.y + v.y);
  
  def addInPlace(self, v):
    self.x += v.x
    self.y += v.y
  
  def scale(self, s):
    return Vector(self.x * s, self.y * s);
  
  def dot(self, v):
    return self.x * v.x + self.y * v.y
  
  def normalize(self):
    magnitude = math.sqrt(self.dot(self))
    if magnitude == 0: return Vector(0, 0)
    return self.scale(1.0 / magnitude)
  
  def __str__(self):
    return "(" + str(self.x) + ", " + str(self.y) + ")"
